src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -625,6 +625,11 @@ public Claim GetClaim(string key)
             return Payload.GetClaim(key, Issuer ?? ClaimsIdentity.DefaultIssuer);
         }
 
+        /// <summary>
+        /// Gets the names of the payload claims on the JsonWebToken.
+        /// </summary>
+        internal IReadOnlyCollection<string> PayloadClaimNames => Payload._jsonClaims.Keys;
```

Comment:
The `Keys` property on a dictionary allocates a `KeyCollection` object.

Reply:
Suggestions for how we can do better?

Reply:
I can't really think of anything. Dictionaries don't expose enough (to protect internal state). I just wanted to call this out in case it was overlooked and maybe someone else might have an idea of how to get the same info via an existing collection.

Reply:
I don't think we can do much better with our current data structures. It is internal now so we can change it in the future if needed

Reply:
It creates a [KeyCollection](https://github.com/dotnet/runtime/blob/284aeaf17c2a1c9eb2cbfc5d3d82f312edbb850a/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L216) and passes dictionary reference, which is what's enumerated over. I think this works?

Another option would be iterating over _jsonClaims directly. Add something like `internal HasClaimType(string type)` to JsonWebToken.